### PR TITLE
Kuyawa merge

### PR DESCRIPTION
**Pull request by kuyawa**
Finished first round of big dev merge

### CODE REVIEW

File: lib/utils/api.ts [hunk 1]
```diff
@@ -1,43 +1,105 @@
...
-    body: JSON.stringify(rec)
+  const result = await postApi('user', rec)
+  console.log('API SAVED USER', result)
+  if(!result || result?.error){
+    return null
+  }
+  return result?.data || null
+}
+
+export async function addAnonymousUser(address:string, chain:string){
```
Comment (JulianKingman) [inline lib/utils/api.ts:61]: This looks like a duplicate function (I see an anonymous.tsx file)

File: lib/utils/api.ts [hunk 2]
```diff
@@ -1,43 +1,105 @@
...
-  const opts = {
-    method: 'POST',
-    headers: { 'Content-Type': 'application/json' },
-    body: JSON.stringify(rec)
-  }
-  const result = await fetch('/api/session', opts)
-  const info   = await result.json()
-  console.log('API SAVED SESSION', info)
-  if(!info || info?.error){
+  const result = await postApi('session', rec)
```
Comment (JulianKingman) [inline lib/utils/api.ts:99]: What does the session API do? I should probably know this, but it's my first time noticing

File: lib/utils/receipt.tsx [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { postApi } from '@/lib/utils/api'
+
+export default async function sendReceipt(data:any){
```
Comment (JulianKingman) [inline lib/utils/receipt.tsx:3]: we should rename the other `sendReceipt` function to `sendEmailReceipt`

File: package.json [hunk 4]
```diff
@@ -49,14 +50,16 @@
     "react": "^18",
     "react-dom": "^18",
     "react-hook-form": "^7.50.0",
+    "react-toastify": "^10.0.5",
```
Comment (JulianKingman) [inline package.json:53]: we can get rid of this

File: package.json [hunk 5]
```diff
@@ -34,8 +34,9 @@
     "@tanstack/react-table": "^8.11.2",
     "@walletconnect/sign-client": "^2.10.6",
     "class-variance-authority": "^0.7.0",
+    "clipboard-copy": "^4.0.1",
```
Comment (JulianKingman) [inline package.json:37]: do we need this, or can we use native web clipboard functions? i.e.
```
await navigator.clipboard.writeText(text);
await navigator.clipboard.readText();
```

File: package-lock.json
Comment (JulianKingman) [inline package-lock.json:1]: delete, use yarn

**Review by JulianKingman — CHANGES_REQUESTED** [review]
Looking really good George! Couple changes.

File: app/api/route.tsx [hunk 6]
```diff
@@ -1,14 +1,31 @@
 //import { searchInitiatives } from '@/lib/utils/registry'
-import { searchOrganizations } from '@/lib/utils/registry'
+import { searchOrganizations, getOrganizations, getStories, getUsers } from '@/lib/utils/registry'
+import { Organization, Story, User } from "@/types/models"
+
+async function testOrganizations():Promise<Organization[]> {
```
[the remark this answers is not in the document]
Comment (kuyawa) [reply app/api/route.tsx:5]: Just testing, will remove them

File: app/initiatives/[id]/page.tsx [hunk 7]
```diff
@@ -82,7 +76,7 @@ export default async function Home(props: any) {
                 See more initiatives 
               </Link>
             }
-            {stories?.length>0 &&
+            {stories?.length &&
```
[the remark this answers is not in the document]
Comment (kuyawa) [reply app/initiatives/[id]/page.tsx:79]: Tell that to the smart bunny

File: app/profile/[userId]/page.tsx [hunk 8]
```diff
@@ -8,25 +8,24 @@ import { getUserById, getNFTsByAccount, getDonationsByUser, getFavoriteOrganizat
...
-  const receipts:[Dictionary]  = await getNFTsByAccount(userid) || []
-  const donations:[Dictionary] = await getDonationsByUser(userid) || []
-  const favorgs:[Dictionary]   = await getFavoriteOrganizations(userid) || []
-  const badges:[Dictionary]    = await getUserBadges(userid) || []
-  const stories:[Dictionary]   = await getRecentStories(5) || []
+  const receipts:[NFTData] = await getNFTsByAccount(userid) || []
+  const donations:[Donation] = await getDonationsByUser(userid) || []
+  const favorgs:[Organization] = await getFavoriteOrganizations(userid) || []
+  const badges:[Category] = await getUserBadges(userid) || []
+  const stories:[Story] = await getRecentStories(5) || []
```
[the remark this answers is not in the document]
Comment (kuyawa) [reply app/profile/[userId]/page.tsx:28]: Fixed

File: components/DonationForm.tsx [hunk 9]
```diff
@@ -1,95 +1,161 @@
...
+import { Initiative, Wallet } from '@/types/models'
 
-export default function DonationForm(props:any) {
+interface IForm {
+  amount: string
+  email: string
+  receipt: boolean
+}
+
+interface IDonation {
```
[the remark this answers is not in the document]
Comment (kuyawa) [reply components/DonationForm.tsx:29]: No, they're temporary data collected to process the donation and payment functions, they will be part of the donation record later on once we add more data

File: components/DonationFormSelect.tsx [hunk 10]
```diff
@@ -56,8 +59,7 @@ const DonationFormSelect = React.forwardRef<HTMLInputElement, SelectInputProps>(
                 key={option.value}
               >
                 <div className="flex flex-row gap-3">
-                  {/* TODO: FIX: IMAGE NOT FOUND IS CAUSING MULTIPLE PAGE RELOADS WITH COIN ID AS INIT ID */}
-                  <img src={option.image} width="30px" />
+                  <Image src={option.image} width={30} height={30} alt="Selected image" onError={(e) => e.currentTarget.style.display = 'none'} />
```
[the remark this answers is not in the document]
Comment (kuyawa) [reply components/DonationFormSelect.tsx:62]: These changes were suggested by the stupid bunny. To be honest I don't like trapping errors in html but if we don't find the image for any reason we should show a placeholder. So, image should be checked for no empty string but error should be trapped if media storage errors out serving the image

File: components/ShareModal.tsx [hunk 11]
```diff
@@ -5,7 +5,10 @@ import Link from "next/link"
 import { useEffect, useState } from 'react'
 import { Command, CommandEmpty, CommandGroup, CommandInput, CommandItem } from '@/components/ui/command'
 import { Popover, PopoverContent, PopoverTrigger } from '@/components/ui/popover'
-import { Share as IconShare } from 'lucide-react'
+import { Share as IconShare, Facebook, Instagram, Twitter, Send as Telegram, MessageCircle as Whatsapp } from 'lucide-react'
+import copy from 'clipboard-copy'
+import { toast, ToastContainer } from 'react-toastify'
```
[the remark this answers is not in the document]
Comment (kuyawa) [reply components/ShareModal.tsx:10]: Again smart bunny is not that smart. I consider the toast too much for such a small UI, we may show a COPIED text in the button instead

File: components/ui/claim-button.tsx [hunk 12]
```diff
@@ -0,0 +1,61 @@
+import * as React from 'react'
+import { Slot } from '@radix-ui/react-slot'
+import { cva, type VariantProps } from 'class-variance-authority'
+import { cn } from '@/lib/shadCnUtil'
+import { ReceiptStatus } from '@/types/common'
+
+export interface ClaimButtonProps extends React.HTMLAttributes<HTMLDivElement> {
+  status: string
```
[the remark this answers is not in the document]
Comment (kuyawa) [reply components/ui/claim-button.tsx:8]: You are right, ReceiptStatus should be it

File: lib/chains/client/base.ts [hunk 13]
```diff
@@ -60,16 +60,23 @@ class BaseSDK{
...
-    })
+    try {
+      this.connect(async (data) => {
+        console.log('Pay connect', data)
+        const result = await this.wallet.payment(address, amount, destinTag)
+        callback(result)
+      })
+    } catch(ex:any) {
+      console.error(ex)
+      callback({error:ex.message})
```
[the remark this answers is not in the document]
Comment (kuyawa) [reply lib/chains/client/base.ts:78]: Or simply use ex:Error?

File: lib/utils/api.ts (hunk 1 above)
Comment (kuyawa) [reply lib/utils/api.ts:61]: Client and server calls. I'll try to merge them but really don't know how to call registry functions from client except from api

Comment (kuyawa) [reply lib/utils/api.ts:61]: Ok, merged into one

File: lib/utils/api.ts (hunk 2 above)
Comment (kuyawa) [reply lib/utils/api.ts:99]: Relic from the past, will remove it

File: lib/utils/receipt.tsx (hunk 3 above)
Comment (kuyawa) [reply lib/utils/receipt.tsx:3]: We may move this to the api lib as it relates to api calls

Comment (kuyawa) [reply lib/utils/receipt.tsx:3]: Done as emailReceipt as it's the only thing it does

File: package.json (hunk 5 above)
Comment (kuyawa) [reply package.json:37]: It was using browser functions before but the clever cottontail suggested to use react libs

File: package.json (hunk 4 above)
Comment (kuyawa) [reply package.json:53]: Removed

File: package-lock.json
Comment (kuyawa) [reply package-lock.json:1]: Done, but I use npm as yarn gives me headaches and bun errors out

**Review by kuyawa — COMMENTED** [review]
All issues solved

**Review by JulianKingman — APPROVED** [review]
I made one change to the claim-button, have a look how I made the types. I first made ReceiptStatus a strict const (not sure if that's the right word) by doing:
```jsx
const ReceiptStatus = {...} as const;
```
I could then use this to create an enum
```jsx
type ReceiptStatusEnum = (typeof ReceiptStatus)[keyof typeof ReceiptStatus]
```
This gets the values of the statuses from the strict const object

**Pull request merged**